test/unit/io/structure_file/structure_file_input_test.cpp

```diff
@@ -0,0 +1,471 @@
+    EXPECT_TRUE((std::is_same_v<typename t::selected_field_ids, comp1>));
+    EXPECT_TRUE((std::is_same_v<typename t::valid_formats,      comp2>));
+    EXPECT_TRUE((std::is_same_v<typename t::stream_type,        comp3>));
+}
+
+TEST_F(structure_file_input_class, modified_traits)
+{
+    test::tmp_filename filename = create_file("> ID\nACGU\n....\n");
+    //! [structure_file_input_class mod_traits]
+    struct my_traits : structure_file_input_default_traits_rna
+    {
+        using seq_alphabet = rna4; // instead of rna5
+    };
+
+    structure_file_in<my_traits> fin{filename.get_path()};
+    //! [structure_file_input_class mod_traits]
+
+    using t = decltype(fin);
+    EXPECT_TRUE((std::is_same_v<typename t::traits_type,        my_traits>));
+    EXPECT_TRUE((std::is_same_v<typename t::selected_field_ids, comp1>));
+    EXPECT_TRUE((std::is_same_v<typename t::valid_formats,      comp2>));
+    EXPECT_TRUE((std::is_same_v<typename t::stream_type,        comp3>));
+}
+
+struct structure_file_input_read : public ::testing::Test
+{
+    size_t const num_records = 2ul;
+
+    std::string const input
+    {
+        ">S.cerevisiae_tRNA-PHE M10740/1-73\n"
+        "GCGGAUUUAGCUCAGUUGGGAGAGCGCCAGACUGAAGAUUUGGAGGUCCUGUGUUCGAUCCACAGAAUUCGCA\n"
+        "(((((((..((((........)))).((((.........)))).....(((((.......)))))))))))). (-17.50)\n"
+        "> example\n"
+        "UUGGAGUACACAACCUGUACACUCUUUC\n"
+        "..(((((..(((...)))..)))))... (-3.71)\n"
+    };
+
+    rna5_vector const seq_comp[2]
+    {
+        "GCGGAUUUAGCUCAGUUGGGAGAGCGCCAGACUGAAGAUUUGGAGGUCCUGUGUUCGAUCCACAGAAUUCGCA"_rna5,
+        "UUGGAGUACACAACCUGUACACUCUUUC"_rna5
+    };
+
+    std::string const id_comp[2]
+    {
+        "S.cerevisiae_tRNA-PHE M10740/1-73",
+        "example"
+    };
+
+    double const energy_comp[2]
+    {
+        -17.5, -3.71
+    };
+
+    std::vector<wuss51> const structure_comp[2]
+    {
+        "(((((((..((((........)))).((((.........)))).....(((((.......))))))))))))."_wuss51,
+        "..(((((..(((...)))..)))))..."_wuss51
+    };
+
+    std::vector<uint8_t> const interaction_comp[2]
+    {
+        {
+            71, 70, 69, 68, 67, 66, 65, 24, 23, 22, 21, 12, 11, 10,  9, 42, 41, 40, 39, 29,
+            28, 27, 26, 64, 63, 62, 61, 60, 52, 51, 50, 49, 48,  6,  5,  4,  3,  2,  1,  0
+        },
+        {
+            24, 23, 22, 21, 20, 17, 16, 15, 11, 10,  9,  6,  5,  4,  3,  2
+        }
+    };
+
+    template <typename bpp_type>
+    void bpp_test(bpp_type & bpp, std::vector<uint8_t> const & bpp_comp)
+    {
+        size_t idx = 0ul;
+        auto interactions = bpp | ranges::view::remove_if([] (auto & set) { return set.size() != 1; });
+        for (auto & elem : interactions)
+        {
+            EXPECT_EQ(elem.begin()->second, bpp_comp[idx++]);
+        }
+        EXPECT_EQ(idx, bpp_comp.size());
+    }
+};
+
+struct structure_file_input_record_reading : public structure_file_input_read {};
```

Comment:
why this?

Reply:
I wanted to create a different test group that has access to the same struct. But I can remove this and rename the tests, if it looks weird.

Reply:
It's not important, really.